File: Assemblies/ProteanCMS/CMS/Cms.xForm.vb
Imports System
Imports System.Xml

Partial Public Class Cms
    Public Class xForm
        Inherits Protean.xForm

#Region "New Error Handling"

        Public Event OnError(ByVal sender As Object, ByVal e As Protean.Tools.Errors.ErrorEventArgs)

        Private Sub _OnError(ByVal sender As Object, ByVal e As Protean.Tools.Errors.ErrorEventArgs) Handles Me.OnError
            returnException(e.ModuleName, e.ProcedureName, e.Exception, "", e.AddtionalInformation, gbDebug)
        End Sub

#End Region

#Region "Declarations"
        Public myWeb As Protean.Cms
        Private mcModuleName As String = "Web.xForm"
#End Region
        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(ByRef aWeb As Protean.Cms)
            MyBase.New()
            PerfMon.Log(mcModuleName, "New")
            Try

                myWeb = aWeb
                moPageXml = myWeb.moPageXml
                mnUserId = myWeb.mnUserId
                moCtx = myWeb.moCtx

                If Not moCtx Is Nothing Then
                    goApp = moCtx.Application
                    goRequest = moCtx.Request
                    goResponse = moCtx.Response
                    goSession = moCtx.Session
                    goServer = moCtx.Server
                End If

            Catch ex As Exception
                RaiseEvent OnError(Me, New Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "New", ex, ""))
            End Try
        End Sub


        Overrides Function evaluateByType(ByVal sValue As String, ByVal sType As String, Optional ByVal cExtensions As String = "", Optional ByVal isRequired As Boolean = False) As String
            Dim cProcessInfo As String = ""
            Dim cReturn As String = "" ' Set this as a clear return string

            Try
                cReturn = MyBase.evaluateByType(sValue, sType, cExtensions, isRequired)

                ' Only evaulate if there is data to evaluate against!

                If Not myWeb Is Nothing And sValue <> "" Then
                    Select Case sType
                        Case "reValidateUser"
                            cReturn = myWeb.moDbHelper.validateUser(mnUserId, sValue)
                            If IsNumeric(cReturn) Then cReturn = ""
                    End Select
                End If

                Return cReturn

            Catch ex As Exception
                RaiseEvent OnError(Me, New Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "evaluteByType", ex, ""))
                Return ""
            End Try
        End Function


        Overrides Function isUnique(ByVal sValue As String, ByVal sPath As String) As Boolean
            Dim cProcessInfo As String = ""
            'Placeholder for overide
            Try

                'Confirm the contenttype and the field
                Dim unqiueVal() As String = sPath.Split("|")
                Dim tableName As String = unqiueVal(0)
                Dim schemaName As String = unqiueVal(1)
                Dim fieldName As String = unqiueVal(2)
                Dim xPath As String = unqiueVal(3)
                'Generate the xpath if value is in XML within the field


                'Query the database to confirm if this value is unique.

                Return myWeb.moDbHelper.isUnique(tableName, schemaName, fieldName, xPath, sValue)


            Catch ex As Exception
                returnException(mcModuleName, "isUnique", ex, "", cProcessInfo, gbDebug)
                Return ""
            End Try
        End Function

    End Class
End Class

File: Assemblies/ProteanCMS/CMS/Cms.xForm.cs
using System;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

using static Protean.stdTools;

namespace Protean
{

    public partial class Cms
    {
        public class xForm : Protean.xForm
        {

            #region New Error Handling

            public event OnErrorEventHandler OnError;

            public delegate void OnErrorEventHandler(object sender, Tools.Errors.ErrorEventArgs e);

            private void _OnError(object sender, Tools.Errors.ErrorEventArgs e)
            {
                stdTools.returnException(ref Protean.xForm.msException, e.ModuleName, e.ProcedureName, e.Exception, "", e.AddtionalInformation, gbDebug);
            }

            #endregion

            #region Declarations
            public Cms myWeb;
            private string mcModuleName = "Web.xForm";
            #endregion
            public xForm(ref string sException) : base(ref sException)
            {
                OnError += _OnError;
            }

            public xForm(ref Cms aWeb) : base(ref aWeb.msException)
            {
                aWeb.PerfMon.Log(mcModuleName, "New");
                try
                {

                    myWeb = aWeb;
                    this.moPageXML = myWeb.moPageXml;
                    this.mnUserId = myWeb.mnUserId;
                    this.moCtx = myWeb.moCtx;

                    if (this.moCtx is not null)
                    {
                        this.goApp = this.moCtx.Application;
                        this.goRequest = this.moCtx.Request;
                        this.goResponse = this.moCtx.Response;
                        this.goSession = this.moCtx.Session;
                        this.goServer = this.moCtx.Server;
                    }
                }

                catch (Exception ex)
                {
                    OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(mcModuleName, "New", ex, ""));
                }

                OnError += _OnError;
            }


            public override string evaluateByType(string sValue, string sType, string cExtensions = "", bool isRequired = false)
            {
                string cProcessInfo = "";
                string cReturn = ""; // Set this as a clear return string

                try
                {
                    cReturn = base.evaluateByType(sValue, sType, cExtensions, isRequired);

                    // Only evaulate if there is data to evaluate against!

                    if (myWeb is not null & !string.IsNullOrEmpty(sValue))
                    {
                        switch (sType ?? "")
                        {
                            case "reValidateUser":
                                {
                                    cReturn = myWeb.moDbHelper.validateUser(this.mnUserId, sValue);
                                    if (Information.IsNumeric(cReturn))
                                        cReturn = "";
                                    break;
                                }
                        }
                    }

                    return cReturn;
                }

                catch (Exception ex)
                {
                    OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(mcModuleName, "evaluteByType", ex, ""));
                    return "";
                }
            }


            public override bool isUnique(string sValue, string sPath)
            {
                string cProcessInfo = "";
                // Placeholder for overide
                try
                {

                    // Confirm the contenttype and the field
                    string[] unqiueVal = sPath.Split('|');
                    string tableName = unqiueVal[0];
                    string schemaName = unqiueVal[1];
                    string fieldName = unqiueVal[2];
                    string xPath = unqiueVal[3];
                    // Generate the xpath if value is in XML within the field


                    // Query the database to confirm if this value is unique.

                    return myWeb.moDbHelper.isUnique(tableName, schemaName, fieldName, xPath, sValue);
                }


                catch (Exception ex)
                {
                    stdTools.returnException(ref myWeb.msException, mcModuleName, "isUnique", ex, "", cProcessInfo, gbDebug);
                    return Conversions.ToBoolean("");
                }
            }

        }
    }
}